File: implementations/c/tests/test_vectors.c
/*
 * ============================================================================
 *  _____                                   ____
 * |_   _|_ _ _ __   __ _  __ _ _ __ __ _  / ___| _ __   __ _  ___ ___
 *   | |/ _` | '_ \ / _` |/ _` | '__/ _` | \___ \| '_ \ / _` |/ __/ _ \
 *   | | (_| | | | | (_| | (_| | | | (_| |  ___) | |_) | (_| | (_|  __/
 *   |_|\__,_|_| |_|\__,_|\__, |_|  \__,_| |____/| .__/ \__,_|\___\___|
 *                        |___/                  |_|
 * ============================================================================
 *
 * POCKET+ C Implementation - Reference Test Vectors
 *
 * Tests compression against reference test vectors to ensure correctness
 * and interoperability with other implementations.
 *
 * Authors:
 *   Georges Labrèche <georges@tanagraspace.com> — https://georges.fyi
 *   Claude Code (claude-sonnet-4-5-20250929) <noreply@anthropic.com>
 * ============================================================================
 */

#include "pocket_plus.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

/* Test vector paths relative to c/ directory */
#define TEST_VECTORS_DIR "../../test-vectors"

/* Test counter */
static int tests_run = 0;
static int tests_passed = 0;
static int tests_failed = 0;

#define TEST(name) \
    static void name(void); \
    static void run_##name(void) { \
        printf("  %s...", #name); \
        fflush(stdout); \
        name(); \
        tests_passed++; \
        printf(" ✓\n"); \
    } \
    static void name(void)

#define RUN_TEST(name) do { \
    tests_run++; \
    run_##name(); \
} while(0)

/* Helper: Read file into buffer */
static size_t read_file(const char *path, uint8_t *buffer, size_t max_size) {
    FILE *fp = fopen(path, "rb");
    if (!fp) {
        fprintf(stderr, "\n  ERROR: Could not open file: %s\n", path);
        return 0;
    }

    fseek(fp, 0, SEEK_END);
    long file_size = ftell(fp);
    rewind(fp);

    if (file_size > (long)max_size) {
        fprintf(stderr, "\n  ERROR: File too large: %ld bytes (max %zu)\n",
                file_size, max_size);
        fclose(fp);
        return 0;
    }

    size_t bytes_read = fread(buffer, 1, file_size, fp);
    fclose(fp);

    return bytes_read;
}

/* Helper: Compare our output with expected output */
static int compare_output(const uint8_t *actual, size_t actual_size,
                         const uint8_t *expected, size_t expected_size,
                         const char *test_name) {
    if (actual_size != expected_size) {
        fprintf(stderr, "\n  FAIL: Size mismatch in %s\n", test_name);
        fprintf(stderr, "    Expected: %zu bytes\n", expected_size);
        fprintf(stderr, "    Got:      %zu bytes\n", actual_size);
        return 0;
    }

    for (size_t i = 0; i < actual_size; i++) {
        if (actual[i] != expected[i]) {
            fprintf(stderr, "\n  FAIL: Byte mismatch in %s at offset %zu\n",
                    test_name, i);
            fprintf(stderr, "    Expected: 0x%02X\n", expected[i]);
            fprintf(stderr, "    Got:      0x%02X\n", actual[i]);

            /* Show some context */
            fprintf(stderr, "    Context (offset %zu):\n", i);
            size_t start = (i > 4) ? i - 4 : 0;
            size_t end = (i + 5 < actual_size) ? i + 5 : actual_size;

            fprintf(stderr, "      Expected: ");
            for (size_t j = start; j < end; j++) {
                fprintf(stderr, "%02X ", expected[j]);
                if (j == i) fprintf(stderr, "[");
            }
            fprintf(stderr, "\n");

            fprintf(stderr, "      Got:      ");
            for (size_t j = start; j < end; j++) {
                fprintf(stderr, "%02X ", actual[j]);
                if (j == i) fprintf(stderr, "[");
            }
            fprintf(stderr, "\n");

            return 0;
        }
    }

    return 1;
}

/* Helper: Compress entire file and compare with expected output */
static int compress_and_verify(
    const char *input_path,
    const char *expected_path,
    size_t packet_length_bits,
    int pt_limit,
    int ft_limit,
    int rt_limit,
    uint8_t robustness,
    const char *test_name
) {
    /* Allocate large buffers for test vectors (venus-express is ~14 MB) */
    static uint8_t input_data[15 * 1024 * 1024];      /* 15 MB for input */
    static uint8_t expected_output[10 * 1024 * 1024]; /* 10 MB for expected output */
    static uint8_t actual_output[10 * 1024 * 1024];   /* 10 MB for actual output */

    /* Read input file */
    size_t input_size = read_file(input_path, input_data, sizeof(input_data));
    if (input_size == 0) {
        fprintf(stderr, "\n  FAIL: Could not read input file: %s\n", input_path);
        return 0;
    }

    /* Read expected output */
    size_t expected_size = read_file(expected_path, expected_output, sizeof(expected_output));
    if (expected_size == 0) {
        fprintf(stderr, "\n  FAIL: Could not read expected output: %s\n", expected_path);
        return 0;
    }

    /* Initialize compressor with automatic parameter management */
    pocket_compressor_t comp;
    int result = pocket_compressor_init(&comp, packet_length_bits, NULL, robustness,
                                        pt_limit, ft_limit, rt_limit);
    if (result != POCKET_OK) {
        fprintf(stderr, "\n  FAIL: Compressor init failed\n");
        return 0;
    }

    /* Compress entire input using high-level API */
    size_t actual_size = 0;

    result = pocket_compress(&comp, input_data, input_size,
                            actual_output, sizeof(actual_output), &actual_size);
    if (result != POCKET_OK) {
        fprintf(stderr, "\n  FAIL: Compression failed with error %d\n", result);
        return 0;
    }

    printf("    Input: %zu bytes → Compressed: %zu bytes (ratio: %.2fx)\n",
           input_size, actual_size, (float)input_size / actual_size);

    /* Compare with expected output */
    return compare_output(actual_output, actual_size,
                         expected_output, expected_size, test_name);
}

/* ========================================================================
 * Test Vector: Simple
 *
 * Parameters from metadata.json:
 *   packet_length: 90 bytes (720 bits)
 *   pt: 10 (new mask every 10 packets)
 *   ft: 20 (send mask every 20 packets)
 *   rt: 50 (uncompressed every 50 packets)
 *   robustness: 1
 * ======================================================================== */

TEST(test_vector_simple) {
    char input_path[256];
    char expected_path[256];
    snprintf(input_path, sizeof(input_path), "%s/input/simple.bin", TEST_VECTORS_DIR);
    snprintf(expected_path, sizeof(expected_path), "%s/expected-output/simple.bin.pkt", TEST_VECTORS_DIR);

    if (!compress_and_verify(input_path, expected_path, 720, 10, 20, 50, 1, "simple")) {
        tests_failed++;
    }
}

/* ========================================================================
 * Test Vector: Housekeeping
 *
 * Parameters from housekeeping.yaml:
 *   packet_length: 90 bytes (720 bits)
 *   pt: 20 (new mask every 20 packets)
 *   ft: 50 (send mask every 50 packets)
 *   rt: 100 (uncompressed every 100 packets)
 *   robustness: 2
 * ======================================================================== */

TEST(test_vector_housekeeping) {
    char input_path[256];
    char expected_path[256];
    snprintf(input_path, sizeof(input_path), "%s/input/housekeeping.bin", TEST_VECTORS_DIR);
    snprintf(expected_path, sizeof(expected_path), "%s/expected-output/housekeeping.bin.pkt", TEST_VECTORS_DIR);

    if (!compress_and_verify(input_path, expected_path, 720, 20, 50, 100, 2, "housekeeping")) {
        tests_failed++;
    }
}

/* ========================================================================
 * Test Vector: Edge Cases
 *
 * Parameters from edge-cases.yaml:
 *   packet_length: 90 bytes (720 bits)
 *   pt: 10 (new mask every 10 packets)
 *   ft: 20 (send mask every 20 packets)
 *   rt: 50 (uncompressed every 50 packets)
 *   robustness: 1
 * ======================================================================== */

TEST(test_vector_edge_cases) {
    char input_path[256];
    char expected_path[256];
    snprintf(input_path, sizeof(input_path), "%s/input/edge-cases.bin", TEST_VECTORS_DIR);
    snprintf(expected_path, sizeof(expected_path), "%s/expected-output/edge-cases.bin.pkt", TEST_VECTORS_DIR);

    if (!compress_and_verify(input_path, expected_path, 720, 10, 20, 50, 1, "edge-cases")) {
        tests_failed++;
    }
}

/* ========================================================================
 * Test Vector: Venus Express
 *
 * Parameters from venus-express.yaml:
 *   packet_length: 90 bytes (720 bits)
 *   pt: 20 (new mask every 20 packets)
 *   ft: 50 (send mask every 50 packets)
 *   rt: 100 (uncompressed every 100 packets)
 *   robustness: 2
 * ======================================================================== */

TEST(test_vector_venus_express) {
    char input_path[256];
    char expected_path[256];
    snprintf(input_path, sizeof(input_path), "%s/input/venus-express.ccsds", TEST_VECTORS_DIR);
    snprintf(expected_path, sizeof(expected_path), "%s/expected-output/venus-express.ccsds.pkt", TEST_VECTORS_DIR);

    if (!compress_and_verify(input_path, expected_path, 720, 20, 50, 100, 2, "venus-express")) {
        tests_failed++;
    }
}

/* ========================================================================
 * Main Test Runner
 * ======================================================================== */

int main(void) {
    printf("\nReference Test Vectors\n");
    printf("======================\n\n");

    printf("Testing against reference implementations...\n");
    printf("Test vectors location: %s\n\n", TEST_VECTORS_DIR);

    RUN_TEST(test_vector_simple);
    RUN_TEST(test_vector_housekeeping);
    RUN_TEST(test_vector_edge_cases);
    RUN_TEST(test_vector_venus_express);

    printf("\n");
    printf("Results: %d/%d tests passed", tests_passed, tests_run);
    if (tests_failed > 0) {
        printf(", %d failed", tests_failed);
    }
    printf("\n\n");

    if (tests_failed > 0) {
        printf("FAIL: Some tests failed unexpectedly.\n\n");
        return 1;
    }

    printf("All tests passed!\n\n");
    return 0;
}

File: implementations/python/tests/test_vectors.py
"""Test vectors validation against reference implementation.

Validates Python POCKET+ output matches byte-for-byte with
expected output from ESA reference implementation.

Run all tests:
    pytest tests/test_vectors.py

Skip slow tests (venus-express, housekeeping):
    pytest tests/test_vectors.py -m "not slow"
"""

import json
from pathlib import Path

import pytest

from pocketplus import compress, decompress

# Path to test vectors (relative to repo root)
REPO_ROOT = Path(__file__).parent.parent.parent.parent
TEST_VECTORS_DIR = REPO_ROOT / "test-vectors"
INPUT_DIR = TEST_VECTORS_DIR / "input"
EXPECTED_DIR = TEST_VECTORS_DIR / "expected-output"

# Large vectors that take significant time to process
SLOW_VECTORS = {"venus-express", "housekeeping"}


def load_metadata(name: str) -> dict:
    """Load metadata JSON for a test vector."""
    metadata_path = EXPECTED_DIR / f"{name}-metadata.json"
    with open(metadata_path) as f:
        return json.load(f)


def get_test_vectors() -> list:
    """Get list of available test vectors with their parameters."""
    vectors = []

    # Map metadata input filenames to actual filenames (some were renamed)
    filename_map = {
        "1028packets.ccsds": "venus-express.ccsds",
    }

    # Find all metadata files
    for metadata_file in EXPECTED_DIR.glob("*-metadata.json"):
        name = metadata_file.stem.replace("-metadata", "")
        metadata = load_metadata(name)

        # Get input filename, applying remap if needed
        input_file = metadata["input"]["file"]
        input_file = filename_map.get(input_file, input_file)

        # Skip if input file doesn't exist
        if not (INPUT_DIR / input_file).exists():
            continue

        vectors.append(
            {
                "name": name,
                "input_file": input_file,
                "output_file": metadata["output"]["compressed"]["file"],
                "packet_length": metadata["compression"]["packet_length"],
                "robustness": metadata["compression"]["parameters"]["robustness"],
                "pt": metadata["compression"]["parameters"]["pt"],
                "ft": metadata["compression"]["parameters"]["ft"],
                "rt": metadata["compression"]["parameters"]["rt"],
                "expected_size": metadata["output"]["compressed"]["size"],
                "expected_md5": metadata["output"]["compressed"]["md5"],
            }
        )

    return vectors


# Get test vectors for parametrization
TEST_VECTORS = get_test_vectors()


def get_parametrized_vectors():
    """Get test vectors with slow markers applied."""
    params = []
    for v in TEST_VECTORS:
        if v["name"] in SLOW_VECTORS:
            params.append(pytest.param(v, marks=pytest.mark.slow, id=v["name"]))
        else:
            params.append(pytest.param(v, id=v["name"]))
    return params


PARAMETRIZED_VECTORS = get_parametrized_vectors()


@pytest.fixture
def vector_data(request):
    """Load input and expected output for a test vector."""
    vector = request.param

    input_path = INPUT_DIR / vector["input_file"]
    expected_path = EXPECTED_DIR / vector["output_file"]

    with open(input_path, "rb") as f:
        input_data = f.read()

    with open(expected_path, "rb") as f:
        expected_output = f.read()

    return {
        **vector,
        "input_data": input_data,
        "expected_output": expected_output,
    }


class TestVectorCompression:
    """Test compression output matches reference."""

    @pytest.mark.parametrize("vector_data", PARAMETRIZED_VECTORS, indirect=True)
    def test_compression_matches_reference(self, vector_data: dict) -> None:
        """Test that compression produces byte-identical output."""
        input_data = vector_data["input_data"]
        expected_output = vector_data["expected_output"]

        # Compress using Python implementation
        # packet_length is in bytes, packet_size is in bits
        packet_size_bits = vector_data["packet_length"] * 8

        actual_output = compress(
            input_data,
            packet_size=packet_size_bits,
            robustness=vector_data["robustness"],
            pt_limit=vector_data["pt"],
            ft_limit=vector_data["ft"],
            rt_limit=vector_data["rt"],
        )

        # Compare byte-for-byte
        assert actual_output == expected_output, (
            f"Compression output mismatch for {vector_data['name']}: "
            f"expected {len(expected_output)} bytes, got {len(actual_output)} bytes"
        )


class TestVectorRoundTrip:
    """Test round-trip decompression."""

    @pytest.mark.parametrize("vector_data", PARAMETRIZED_VECTORS, indirect=True)
    def test_round_trip(self, vector_data: dict) -> None:
        """Test that compress then decompress returns original."""
        input_data = vector_data["input_data"]
        packet_size_bits = vector_data["packet_length"] * 8

        # Compress
        compressed = compress(
            input_data,
            packet_size=packet_size_bits,
            robustness=vector_data["robustness"],
            pt_limit=vector_data["pt"],
            ft_limit=vector_data["ft"],
            rt_limit=vector_data["rt"],
        )

        # Decompress
        decompressed = decompress(
            compressed,
            packet_size=packet_size_bits,
            robustness=vector_data["robustness"],
        )

        assert decompressed == input_data, (
            f"Round-trip failed for {vector_data['name']}: "
            f"expected {len(input_data)} bytes, got {len(decompressed)} bytes"
        )


class TestVectorDecompressReference:
    """Test decompression of reference compressed output."""

    @pytest.mark.parametrize("vector_data", PARAMETRIZED_VECTORS, indirect=True)
    def test_decompress_reference_output(self, vector_data: dict) -> None:
        """Test that decompressing reference output produces original input."""
        input_data = vector_data["input_data"]
        expected_output = vector_data["expected_output"]
        packet_size_bits = vector_data["packet_length"] * 8

        # Decompress the reference compressed output
        decompressed = decompress(
            expected_output,
            packet_size=packet_size_bits,
            robustness=vector_data["robustness"],
        )

        assert decompressed == input_data, (
            f"Decompression of reference failed for {vector_data['name']}: "
            f"expected {len(input_data)} bytes, got {len(decompressed)} bytes"
        )
